### PR TITLE
feat: add per-brand price adjustments for cheapest nearby mode

### LINKED ISSUE #304: [Feature Request]: Per-brand price adjustment for "cheapest nearby" mode
### Describe the feature

Now that we have an integration that finds the cheapest gas price nearby, with options to limit to brand or station, it would be neat to be able to modify certain gas station prices by a fixed amount during this determination.

For example, I have a helper template that modifies my local Walmart gas prices, since I get $0.10 discount per gallon. 

Not really sure how this would fit into Home Assistant's app interface though. Potentially an optional step after identifying the included brands?

### Additional information

Related to #290

### PR DESCRIPTION
**Pull request by firstof9**
## Proposed change

Add support for configuring per-brand price adjustments (discounts or markups) in the cheapest gas tracker flow. This allows comparing stations with different brand loyalty discounts or benefits (e.g. Walmart, Costco, etc.), while keeping the reported state of the sensor as the actual retail price at the pump.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #304
- This PR is related to issue:

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] The code has been formatted using Ruff (`ruff format custom_components/gasbuddy tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated in `README.md`

If the code communicates with devices, web services, or third-party tools:

- [ ] The `manifest.json` file has all fields filled out correctly.
- [ ] For the updated dependencies a diff between library versions and ideally a link to the changelog/release notes is added to the PR description.

## Summary by CodeRabbit

* **New Features**
  * Added brand price adjustments for the “cheapest station” tracker, letting you account for brand-specific discounts or markups.
  * The configuration screens now include a dedicated field and a link to setup guidance.

* **Bug Fixes**
  * Improved cheapest-station selection so adjusted prices are used for comparison while the displayed sensor price stays the actual pump price.
  * Added handling for invalid adjustment values and missing deal prices.

* **Documentation**
  * Expanded the README with examples and an explanation of how brand adjustments affect comparisons.

### CODE REVIEW

File: custom_components/gasbuddy/config_flow.py [hunk 1]
```diff
@@ -951,6 +961,9 @@ async def async_step_cheapest_filters(self, user_input=None):
             data_schema=_get_schema_cheapest_filters(
                 self.hass, brands, stations, user_input, self._data
             ),
+            description_placeholders={
+                "brand_adjustments_url": "https://github.com/firstof9/ha-gasbuddy#brand-price-adjustments"
+            },
```
[the remark this answers is not in the document]
Comment (firstof9) [reply custom_components/gasbuddy/config_flow.py:966]: All localization files (`es.json`, `fr.json`, `pt.json`) have been updated to include the new `brand_adjustments` string key and the matching `{brand_adjustments_url}` placeholder description link.

File: custom_components/gasbuddy/coordinator.py [hunk 2]
```diff
@@ -354,16 +356,50 @@ async def _async_update_cheapest(self) -> dict:  # noqa: PLR0914
...
+                    except TypeError:
+                        pass
+            else:
+                for b_name in station_brand_names:
+                    matching_adjustments = [
+                        val
+                        for key, val in brand_adjustments.items()
+                        if str(key).lower() == b_name.lower()
+                        or (b_name.lower().startswith(str(key).lower()) and len(str(key)) > 3)
+                    ]
```
[the remark this answers is not in the document]
Comment (firstof9) [reply custom_components/gasbuddy/coordinator.py:379]: The fallback brand-name lookup has been refactored to require an exact case-insensitive match instead of allowing prefix-based matching. This avoids matching unintended brands and aligns with the documented behavior.

File: custom_components/gasbuddy/strings.json [hunk 3]
```diff
@@ -80,21 +80,23 @@
...
             "reconfigure_cheapest_filters": {
-                "description": "Configure brand and station filters to restrict which stations are tracked.",
+                "description": "Configure brand and station filters to restrict which stations are tracked. See [documentation]({brand_adjustments_url}) for brand price adjustment examples.",
                 "data": {
                     "exclude_brands": "Excluded Brands",
                     "include_brands": "Included Brands",
                     "exclude_stations": "Excluded Stations",
-                    "include_stations": "Included Stations"
+                    "include_stations": "Included Stations",
+                    "brand_adjustments": "Brand Price Adjustments"
```
[the remark this answers is not in the document]
Comment (firstof9) [reply custom_components/gasbuddy/strings.json:99]: Spanish, French, and Portuguese translation files have been updated and synced with the new key and the `{brand_adjustments_url}` placeholder link.

**Review by secondof9 — APPROVED** [review]

**Comment by secondof9** [comment]
# Code Review: PR #305 — Per-Brand Price Adjustments

## Summary
This PR introduces support for per-brand price adjustments (discounts/markups) when determining the cheapest gas station. The adjustments are applied only during cheapest-station comparisons; the sensor still reports the actual retail pump price.

## ✅ Looks Good
- **README.md** — Clear documentation with examples
- **const.py** — Proper constant and version bump
- **__init__.py** — Correct migration from v8→v9
- **config_flow.py** — Proper ObjectSelector integration
- **strings.json / translations** — Proper i18n with doc links
- **tests/test_config_flow.py** — Valid flow test
- **tests/test_sensor.py** — Core functionality tests are well-structured

## ⚠️ Warnings

### 1. Misapplied `else` block in brand matching logic
**File:** `custom_components/gasbuddy/coordinator.py`  **Line:** 383  **Issue:** The `else` block after `for b_id in station_brand_ids:` executes on every station regardless of whether a brandId match was found. The `break` after a successful lookup doesn't prevent the `else` block from running on subsequent stations.

**Suggestion:** Restructure the matching logic to properly handle both brandId and name matching without the broken `else` guard.

### 2. Invalid edge case tests for error handling  **File:** `tests/test_sensor.py`  **Line:** 888  **Issue:** Uses invalid types (`"invalid_float"`, `[]`) to test error handling. The `try/except` blocks catch exceptions and continue with `adjustment = 0.0`, silently ignoring malformed config instead of falling back to name matching.

**Suggestion:** Replace invalid type tests with valid float/string adjustments to properly test error handling.

## 💡 Suggestions
- **coordinator.py:333** — Remove redundant `or {}` default after `.get()`
- **coordinator.py:346-395** — Restructure brand matching logic to properly handle both brandId and name matching without the broken `else` guard
- **coordinator.py:370-383** — Fix the `adjust()` function to not treat `None` as `inf` when fallback prices exist
- **tests/test_sensor.py:888** — Replace invalid type tests with valid float/string adjustments

## 🔒 Security
No security concerns identified.

## ⚡ Performance
List comprehension for brand ID extraction is fine; nested loops for name matching could be optimized with dictionary lookups, but station counts are typically small — not critical.

**Comment by firstof9** [comment]
Thanks for the review and suggestions, @secondof9! 

I have refactored and addressed the feedback:
1. **Simplified Brand matching logic**: Removed the `for...else` construct and replaced it with a cleaner, explicit `matched` boolean flag.
2. **Warn on malformed config**: Added `_LOGGER.warning` logs when catching ValueError/TypeError during adjustment lookup, so malformed config is no longer silently ignored.

All unit tests run successfully, maintaining **100% test coverage**.
